fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/cursors/throttled/ThrottledRetryingIterator.java

```diff
@@ -0,0 +1,595 @@
+            return futureManager.registerFuture(result);
+        } else {
+            return AsyncUtil.READY_TRUE;
+        }
+    }
+
+    private CompletableFuture<Boolean> handleFailure(Throwable ex, QuotaManager quotaManager) {
+        // Note: the transactional runner does not retry internally
+        ++failureRetriesCounter;
+        if (failureRetriesCounter > numOfRetries) {
+            if (logger.isWarnEnabled()) {
+                logger.warn(KeyValueLogMessage.of("ThrottledIterator: iterate one range failure: will abort",
+                        LogMessageKeys.LIMIT, cursorRowsLimit,
+                        LogMessageKeys.RETRY_COUNT, failureRetriesCounter),
+                        ex);
+            }
+
+            // Complete exceptionally
+            return CompletableFuture.failedFuture(ex);
+        }
+        // Here: after a failure, try setting a scan quota that is smaller than the number of scanned items during the failure
+        successCounter = 0;
+        final int oldLimit = cursorRowsLimit;
+        cursorRowsLimit = decreaseLimit(quotaManager.scannedCount);
+        if (logger.isInfoEnabled() && (oldLimit != cursorRowsLimit)) {
+            logger.info(KeyValueLogMessage.of("ThrottledIterator: iterate one range failure: will retry",
+                            LogMessageKeys.LIMIT, cursorRowsLimit,
+                            LogMessageKeys.OLD_LIMIT, oldLimit,
+                            LogMessageKeys.RETRY_COUNT, failureRetriesCounter),
+                    ex);
+        }
+
+        return AsyncUtil.READY_TRUE; // retry
+    }
+
+    @VisibleForTesting
+    static long throttlePerSecGetDelayMillis(long rangeProcessingTimeMillis, int maxPerSec, int eventsCount) {
+        if (maxPerSec <= 0) {
+            return 0; // do not throttle
+        }
+        // get the number of events, get the min time they should have taken,
+        // and return a padding time (if positive)
+        // MS(count / perSec) - ptimeMillis ==>  MS(count) / perSec - ptimeMillis (avoid floating point, the floor effect is a neglectable 0.005%)
+        long waitMillis = (TimeUnit.SECONDS.toMillis(eventsCount) / maxPerSec) - rangeProcessingTimeMillis;
+        return waitMillis > 0 ? waitMillis : 0;
+    }
+
+    private long nowMillis() {
+        return System.currentTimeMillis();
+    }
+
+    private long elapsedTimeMillis() {
+        return rangeIterationStartTimeMilliseconds <= 0 ? 0 :
+               nowMillis() - rangeIterationStartTimeMilliseconds;
+    }
+
+    private static void runUnlessNull(@Nullable Consumer<QuotaManager> func, QuotaManager quotaManager) {
+        if (func != null) {
+            func.accept(quotaManager);
+        }
+    }
+
+    @VisibleForTesting
+    static int increaseLimit(final int current, final int max) {
+        if (current == 0) {
+            return 0;
+        }
+        int newLimit = Math.max((current * 5) / 4, current + 1);
```

Comment:
If the count is under 16 (let alone 4), the assumption is that there was a problematic item. After 40 successes, though, we can probably raise a little more aggressively. Note that the default time limit makes sure that it will break only if the last item takes more than a second to process. 
```suggestion
        int newLimit = Math.max((current * 5) / 4, current + 4);
```

Reply:
Done.